kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopCredentialsManager.scala

```diff
@@ -0,0 +1,245 @@
+  private val remoteRefMap = new ConcurrentHashMap[String, RemoteCredentialsRef]()
+
+  private var providers: Map[String, HadoopDelegationTokenProvider] = _
+  private var renewalInterval: Long = _
+  private var renewalRetryWait: Long = _
+  private var renewalExecutor: ScheduledExecutorService = _
+  private var hadoopConf: Configuration = _
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    hadoopConf = KyuubiHadoopUtils.newHadoopConf(conf)
+    providers = HadoopCredentialsManager.loadProviders(conf)
+      .filter { case (_, provider) =>
+        val required = provider.delegationTokensRequired(hadoopConf, conf)
+        if (!required) {
+          info(s"Service ${provider.serviceName} does not require a token." +
+            s" Check your configuration to see if security is disabled or not.")
+        }
+        required
+      }
+    info("Using the following builtin delegation token providers: " +
+      s"${providers.keys.mkString(", ")}.")
+
+    renewalInterval = conf.get(CREDENTIALS_RENEWAL_INTERVAL)
+    renewalRetryWait = conf.get(CREDENTIALS_RENEWAL_RETRY_WAIT)
+    super.initialize(conf)
+  }
+
+  override def start(): Unit = {
+    renewalExecutor =
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("Delegation Token Renewal Thread")
+    super.start()
+  }
+
+  override def stop(): Unit = {
+    if (renewalExecutor != null) {
+      renewalExecutor.shutdownNow()
+      try {
+        renewalExecutor.awaitTermination(10, TimeUnit.SECONDS)
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+    super.stop()
+  }
+
+  /**
+   * Send credentials to SQL engine which the specified session is talking to if
+   * [[HadoopCredentialsManager]] has a newer credentials.
+   * @param sessionId Specify the session which is talking with SQL engine
+   * @param appUser  User identity that the SQL engine uses.
+   * @param send     Function to send encoded credentials to SQL engine
+   * @throws KyuubiException if sending credentials fails.
+   */
+  def sendCredentialsIfNeeded(
+      sessionId: String,
+      appUser: String,
+      send: String => Unit): Unit = {
+    require(renewalExecutor != null, "renewalExecutor should be initialized")
+
+    val userRef = getOrCreateUserRef(appUser)
+    val engineRef = getOrCreateEngineRef(sessionId)
+    if (userRef.getEpoch != engineRef.getEpoch) {
+      engineRef synchronized {
+        if (userRef.getEpoch != engineRef.getEpoch) {
+          val currentEpoch = userRef.getEpoch
+          val currentCreds = userRef.getEncodedCredentials
+          info(s"Send new credentials with epoch $currentEpoch to SQL engine through session " +
+            s"$sessionId")
+          Try(send(currentCreds)) match {
+            case Success(_) =>
+              info(s"Update session scope epoch from ${engineRef.getEpoch} to $currentEpoch")
+              engineRef.setEpoch(currentEpoch)
+            case Failure(exception) =>
+              throw new KyuubiException(
```

Comment:
How do we handle this exception later?

Reply:
Since method caller can do nothing but record the exception, changed to record exception in current method.